src/libtls/tls_server.c

```diff
@@ -240,29 +314,58 @@ static status_t process_client_hello(private_tls_server_t *this,
 		extensions = bio_reader_create(ext);
 		while (extensions->remaining(extensions))
 		{
-			if (!extensions->read_uint16(extensions, &extension) ||
-				!extensions->read_data16(extensions, &ext))
+			if (!extensions->read_uint16(extensions, &extension_type) ||
+				!extensions->read_data16(extensions, &extension_data))
 			{
 				DBG1(DBG_TLS, "received invalid ClientHello Extensions");
 				this->alert->add(this->alert, TLS_FATAL, TLS_DECODE_ERROR);
 				extensions->destroy(extensions);
 				return NEED_MORE;
 			}
+			extension = bio_reader_create(extension_data);
 			DBG2(DBG_TLS, "received TLS '%N' extension",
-				 tls_extension_names, extension);
-			DBG3(DBG_TLS, "%B", &ext);
-			switch (extension)
+				 tls_extension_names, extension_type);
+			DBG3(DBG_TLS, "%B", &extension_data);
+			switch (extension_type)
 			{
 				case TLS_EXT_SIGNATURE_ALGORITHMS:
-					this->hashsig = chunk_clone(ext);
+					this->hashsig = chunk_clone(extension_data);
 					break;
 				case TLS_EXT_SUPPORTED_GROUPS:
 					this->curves_received = TRUE;
-					this->curves = chunk_clone(ext);
+					this->curves = chunk_clone(extension_data);
+					break;
+				case TLS_EXT_SUPPORTED_VERSIONS:
+					if (!extension->read_data8(extension, &versions))
+					{
+						DBG1(DBG_TLS, "invalid %N extension",
+		   					 tls_extension_names, extension_type);
+						this->alert->add(this->alert, TLS_FATAL,
+					   					 TLS_DECODE_ERROR);
+						extensions->destroy(extensions);
+						extension->destroy(extension);
+						return NEED_MORE;
+					}
+					break;
+				case TLS_EXT_KEY_SHARE:
+					if (!extension->read_uint16(extension, &key_share_length) ||
+						!extension->read_uint16(extension, &key_type) ||
+						(!extension->read_data16(extension, &key_share) &&
+						key_share.len))
```

Comment:
I know it's the same as on the client, but this check doesn't look right. It should probably be something like this:
```suggestion
						!extension->read_data16(extension, &key_share) ||
						!key_share.len)
```
I'll fix that on the client too.